anndata/_core/merge.py

```diff
@@ -122,7 +122,11 @@ def equal_dask_array(a, b) -> bool:
     if isinstance(b, DaskArray):
         if tokenize(a) == tokenize(b):
             return True
-    return da.equal(a, b, where=~(da.isnan(a) == da.isnan(b))).all()
+    if isinstance(a._meta, spmatrix):
+        # TODO: Maybe also do this in the other case?
+        return da.map_blocks(equal, a, b, drop_axis=(0, 1)).all()
```

Comment:
Any specific reason to not use np.equal? Wouldn't it check type for each block even though the block type is the same for all?

Reply:
`np.equal` doesn't work on sparse arrays:

```python
In [3]: X = sparse.random(
   ...:     1000,
   ...:     100,
   ...:     format="csr",
   ...:     density=0.01,
   ...:     random_state=np.random.default_rng(),
   ...: )

In [4]: np.equal(X, X)
/usr/local/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3508: SparseEfficiencyWarning: Comparing sparse matrices using == is inefficient, try using != instead.
  exec(code_obj, self.user_global_ns, self.user_ns)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[4], line 1
----> 1 np.equal(X, X)

File /usr/local/lib/python3.9/site-packages/scipy/sparse/_base.py:332, in _spbase.__bool__(self)
    330     return self.nnz != 0
    331 else:
--> 332     raise ValueError("The truth value of an array with more than one "
    333                      "element is ambiguous. Use a.any() or a.all().")

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all().
```

Plus we want the `nan` handling behavior here.

Reply:
True, sorry, I guess I made a mistake because it somehow passed for me when I wrote it, but now it has failed. Then it would make sense if that is just the else. However it both fails for me in this case

```python
def test_sparse_vs_dense_dask():

    data = np.random.random((100, 100))
    x = as_sparse_dask_array(data)
    y = as_dense_dask_array(data)
    assert equal(y,x)
    assert equal(x,y)
```